### PR TITLE
feat(python): much faster lazy type-checks

### PR DESCRIPTION
**Pull request by alexander-beedie**
First-pass fix for a performance regression (ref #6052).

* Makes all of the lazy type-checks ~7-8x faster. 
* Renamed the checks to better-describe what they are actually doing
* Moved the expensive checks last, during initial type-determination (frame/series init).
* Faster `Sequence` type-checks for the common case (~3.5x).

Will do another pass later to take care of other low-hanging fruit.

### CODE REVIEW

File: py-polars/polars/internals/dataframe/frame.py [hunk 1]
```diff
@@ -274,20 +274,20 @@ def __init__(
...
 
-        elif _NUMPY_TYPE(data) and isinstance(data, np.ndarray):
-            self._df = numpy_to_pydf(data, columns=columns, orient=orient)
-
-        elif _PYARROW_TYPE(data) and isinstance(data, pa.Table):
-            self._df = arrow_to_pydf(data, columns=columns)
-
         elif isinstance(data, Sequence) and not isinstance(data, str):
             self._df = sequence_to_pydf(
                 data, columns=columns, orient=orient, infer_schema_length=50
```
Comment (ghuls) [inline py-polars/polars/internals/dataframe/frame.py:279]: Can you also switch the Sequence and Series instance checks?
Sequence check takes 10 times longer than a plain instance check

File: py-polars/polars/dependencies.py [hunk 2]
```diff
@@ -163,25 +165,26 @@ def _lazy_import(module_name: str) -> tuple[ModuleType, bool]:
...
-def _NUMPY_TYPE(obj: Any) -> bool:
-    return _NUMPY_AVAILABLE and any(
-        "numpy." in str(o)
-        for o in (obj if inspect.isclass(obj) else obj.__class__).mro()
-    )
+@cache
+def _might_be(cls: type, type_: str) -> bool:
+    # infer whether the given class "might" be associated with the given
+    # module (in which case it's reasonable to do a real isinstance check)
+    try:
```
Comment (ghuls) [inline py-polars/polars/dependencies.py:172]: Can you use https://github.com/pola-rs/polars/commit/4bd67da070e667b3cb2ee51cff9afa69c825d106 ?
Latest ruff, will flag it.

File: py-polars/polars/internals/dataframe/frame.py [hunk 3]
```diff
@@ -274,20 +274,20 @@ def __init__(
...
+        elif _check_for_numpy(data) and isinstance(data, np.ndarray):
+            self._df = numpy_to_pydf(data, columns=columns, orient=orient)
+
+        elif _check_for_pyarrow(data) and isinstance(data, pa.Table):
+            self._df = arrow_to_pydf(data, columns=columns)
+
+        elif _check_for_pandas(data) and isinstance(data, pd.DataFrame):
             self._df = pandas_to_pydf(data, columns=columns)
 
         elif isinstance(data, (Generator, Iterable)) and not isinstance(data, Sized):
```
Comment (ghuls) [inline py-polars/polars/internals/dataframe/frame.py:293]: Switching the condition here might also be faster (not sure how common it is to have Sixed):
```elif not isinstance(data, Sized) and isinstance(data, (Generator, Iterable))```

File: py-polars/polars/internals/dataframe/frame.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/internals/dataframe/frame.py]: Dropped the unnecessary `and not isinstance(data, str)` that follows the `Sequence` check too. If somebody genuinely wants to init a frame with a string, who am I to stand in their way... ;)
```python
pl.DataFrame("xyz")
# ┌──────────┐
# │ column_0 │
# │ ---      │
# │ str      │
# ╞══════════╡
# │ x        │
# │ y        │
# │ z        │
# └──────────┘
```

File: py-polars/polars/internals/dataframe/frame.py [hunk 4]
```diff
@@ -274,13 +274,13 @@ def __init__(
...
             self._df = dict_to_pydf(data, columns=columns)
 
-        elif isinstance(data, Sequence) and not isinstance(data, str):
-            self._df = sequence_to_pydf(
-                data, columns=columns, orient=orient, infer_schema_length=50
-            )
         elif isinstance(data, pli.Series):
             self._df = series_to_pydf(data, columns=columns)
 
+        elif isinstance(data, (list, tuple, Sequence)):
```
Comment (ghuls) [inline py-polars/polars/internals/dataframe/frame.py:280]: should it have the `not isinstance(data, str)` check or is it enough that it is handled in sequence_to_pydf?

File: py-polars/polars/internals/dataframe/frame.py (hunk 3 above)
Comment (alexander-beedie) [reply py-polars/polars/internals/dataframe/frame.py]: Makes sense to me; done.

File: py-polars/polars/internals/dataframe/frame.py (hunk 4 above)
Comment (alexander-beedie) [reply py-polars/polars/internals/dataframe/frame.py:280]: Turns out it's actually [fine](https://github.com/pola-rs/polars/pull/6064#discussion_r1062412379) to drop it :)

**Comment by jakob-keller** [comment]
An alternative, more extendable pattern could be something along these lines:

```
Data = dict[str, Sequence[Any]] | Sequence[Any] | np.ndarray[Any, Any] | pa.Table | pd.DataFrame | pli.Series | None.  # 1st argument of DataFrame.__init__()
ToPyDF = Callable[[Data, ColumnsType | None, Orientation | None], PyDataFrame]

_AVAILABLE_TO_PYDFS: dict[type, ToPyDF] = {
    type(None): (lambda d, c, o: dict_to_pydf({}, columns=c)),
    dict: (lambda d, c, o: dict_to_pydf(d, columns=c)),
    Sequence: (lambda d, c, o: sequence_to_pydf(d, columns=c, orient=o, infer_schema_length=50)),
    # ...
}
# conditionally add numpy, pandas, ... related entries

@functools.cached
def _get_to_pydf(type_: type) -> ToPyDf:
    try:
        return _AVAILABLE_TO_PYDFS[type_]
    except KeyError:
        for t, to_pydf in _AVAILABLE_TO_PYDFS.items():
            if issubclass(type_, t):
                return to_pydf

        raise

class DataFrame:
    def __init__(self,
        data: Data = None,
        columns: ColumnsType | None = None,
        orient: Orientation | None = None,
    )
        try:
            to_pydf = _get_to_pydf(type(data))
        except KeyError as exc:
            raise ValueError(
                f"DataFrame constructor called with unsupported type; got {type(data)}"
        ) from exc

        self._df = to_pydf(data, columns, orient)
```

**Comment by alexander-beedie** [comment]
> An alternative, more extendable pattern could be something along these lines:

Caching the type for the lazy-check in this PR gave me _extremely_ similar thoughts ;) 

However, I want to separate structural changes from a simple speeding-up of what's already there to keep each commit reasonably targeted. Something along those Iines will be part of a second pass though (I want to make it even more generic by reaching down into the constructor functions like `sequence_to_pydf` and doing much the same thing).

**Pull request merged**